[workloadmeta/docker] Collect SBOMs using Trivy

sblumenthal opened the pull request:
### What does this PR do?

This PR adds support for collecting SBOM data from the docker runtime via trivy.

### Motivation

At the time of authoring this PR, SBOM data is only supported by containerd. However, not every environment or configuration uses this container runtime, so this PR adds support for docker, which should allow it to be used in more environments.

### Additional Notes

- Much of the methodology behind this implementation was adapted from the implementation for containerd. Notably, some internal trivy functionality was copied over.
- Right now, this PR does not support scanning docker images from the local filesystem, unlike the containerd implementation. I am happy to add that here, I am also happy to add that in a follow-up PR.

### Describe how to test/QA your changes

Run the agent locally using docker via `docker run` with the env vars `-e DD_CONTAINER_IMAGE_COLLECTION_METADATA_ENABLED=true -e DD_CONTAINER_IMAGE_COLLECTION_SBOM_ENABLED=true` set, and verify output using `agent workload-list --verbose | grep SBOM`

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

sblumenthal committed:
Add initial working pass at collecting sbom data from docker

sblumenthal committed:
Small refactor to remove reliance on containerd util pkg

- Before this change, any artifact that was built with the trivy build
  flag would also require the containerd build flag, due to an interface
  that was being used. This commit redefines the relevant functions in a
  new, local interface to prevent that additional tag from being required.

vboulineau commented on pkg/util/trivy/docker.go:
```diff
@@ -0,0 +1,59 @@
+		return nil, cleanup, xerrors.Errorf("unable to get history (%s): %w", imageID, err)
+	}
+
+	f, err := os.CreateTemp("", "fanal-*")
```
```suggestion
	f, err := os.CreateTemp("", "fanal-docker-*")
```

vboulineau commented on pkg/workloadmeta/collectors/internal/docker/image_sbom_trivy.go:
```diff
@@ -0,0 +1,162 @@
+		return c.dockerUtil.RawClient(), nil
+	}
+
+	c.trivyClient, err = trivy.NewCollector(trivyConfiguration)
```
Having a new collector implies that we're potentially creating a second Trivy instance, and a second cache instance, with the same parameters. There are at least two things to check:
- No race/deadlock/similar created by accessing the cache file in parallel
- No multiple scans in parallel.

These are the same issues as described in https://github.com/DataDog/datadog-agent/pull/15686 (which will conflict with this PR), though a lot less frequent in this case as it's les likely to have both container runtimes used on a host.

sblumenthal committed:
Address PR comments

sblumenthal committed:
Merge branch 'main' into sblumenthal/sbom-docker

# Conflicts:
#	pkg/util/trivy/trivy.go

sblumenthal committed:
Add disk space check to docker trivy scan

sblumenthal committed:
Remove out of scope trivy containerd client refactor

- A lot has changed in this codebase since this refactor was introduced,
  and while the idea behind it is a good one, to minimize the impact of this
  change I am removing this.

sblumenthal commented:
Closing in favor of https://github.com/DataDog/datadog-agent/pull/16528
